### PR TITLE
Majority of the v1 implementation

### PR DESCRIPTION
**Pull request by shashank2000**
- added dataloaders and preprocessing and the rest of the skeleton
- added contrastive loss implementation in BERT model
- will add mlm loss in the next revision

### CODE REVIEW

File: data_module.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                    transforms.RandomResizedCrop(224, scale=(0.2, 1.)),
+                    transforms.RandomHorizontalFlip(),
+                    transforms.ToTensor(),
+                    transforms.Normalize(mean=[0.491, 0.482, 0.446],
+                            std=[0.247, 0.243, 0.261]),
+                ]
+                )  
+    
+    # read env variables here
+    self.questions_file = os.environ.get('QUESTIONS_FILE')
```
Comment (mhw32) [inline data_module.py]: Might be better to store these constants in config rather than in global variables

File: data_module.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    self.answers_file = os.environ.get('ANSWERS_FILE')
+    self.coco_loc = os.environ.get('COCO_LOC')   
+    
+    saved_train_file = 'train{}_correct_padding.pt'.format(self.is_dumb(dumb_transfer, num_answers, multiple_images))
```
Comment (mhw32) [inline data_module.py]: what is `is_dumb` mean? add comments

File: data_module.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                        num_workers=self.num_workers, pin_memory=True, drop_last=True)
+
+  def test_dataloader(self):
+      return DataLoader(self.test_dataset, batch_size=self.batch_size,
+                        num_workers=self.num_workers, pin_memory=True, drop_last=True)
+
+  def get_vocab_length(self):
+    return self.train_dataset.vocabulary_length()
+
+  @staticmethod
```
Comment (mhw32) [inline data_module.py:137]: add comment

File: dataset/jeopardy_dataset.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+from torch.utils.data import Dataset
+import torch
+import os
+import string
+from tqdm import tqdm
+import nltk
+
+# START_TOKEN = "<s>"
+# END_TOKEN = "</s>"
+PAD_TOKEN = "<pad>"
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: I don't think we need to specify these given the roberta tokenizer anymore? Delete if so

File: dataset/jeopardy_dataset.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+UNK_TOKEN = "<unk>"
+
+class JeopardyDataset(Dataset):
+    def __init__(self, 
+        questions_file, 
+        answers_file, 
+        images_dir, 
+        transform, 
+        word2idx=None,
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: do we need this anymore?

File: dataset/jeopardy_dataset.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+class JeopardyDataset(Dataset):
+    def __init__(self, 
+        questions_file, 
+        answers_file, 
+        images_dir, 
+        transform, 
+        word2idx=None, 
+        train=True, 
+        q_len=8,
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:32]: why 8? default roberta encoding is like 512 if i remember correctly

File: dataset/jeopardy_dataset.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        """
+        self.dumb_transfer = dumb_transfer
+        self.multiple_images = multiple_images
+        # initializing the lengths of our questions and answers
+        self.q_len = q_len
+        self.ans_len = ans_len
+        self.num_answers = num_answers
+        # return test or train set?
+        self.train = train
+        self.glove_dict = self._get_glove_indices()
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: Do we need this? Prune all code that's no longer useful to keep repo clean

File: dataset/jeopardy_dataset.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+        q_f = open(questions_file, 'r')
+        a_f = open(answers_file, 'r')
+        
+        # we want only 80% of the question/answer text to build the vocabulary
+        self.answers = json.load(a_f)["annotations"]
+        self.questions = json.load(q_f)["questions"]
+
+        split_index = int((1-test_split) * len(self.questions))
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:67]: make sure the questions aren't sorted in any special way where if we take the last `test_split` as test , they are all of a certain class or something

File: dataset/jeopardy_dataset.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    def _build_word2idx(self, vocab):
+        '''
+        word2idx[PAD_TOKEN] = len(glove_dict)
+        '''
+        word2idx = {w:self.glove_dict[w] for w in vocab}
+        word2idx[UNK_TOKEN] = len(self.glove_dict) - 2 #sandberger is the word for unknown, we learn this embedding over time
+        word2idx[PAD_TOKEN] = len(self.glove_dict) - 1
+        return word2idx
+
+    def _words_to_indices(self, sentence_array, answer=False):
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: remove

File: dataset/jeopardy_dataset.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+            # will always be set to UNK_TOKEN if its more than a word long
+            # what should I be doing here
+            # its just a word now, "array" is misleading
+            sentence_array = sentence_array.lower()
+            if sentence_array not in self.word2idx:
+                sentence_array = UNK_TOKEN
+            return self.word2idx[sentence_array]
+        return [self.word2idx[word] if word in self.word2idx else self.word2idx[UNK_TOKEN] for word in sentence_array]
+
+    def _make_vocab(self):
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: remove

File: dataset/jeopardy_dataset.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        vocab = [word.lower() for word in vocab]
+        copy_vocab = vocab.copy()
+        for word in copy_vocab:
+            # remove everything not in glove TODO: re-run Jeopardy with this
+            if word not in self.glove_dict:
+                vocab.remove(word)
+        self.vocab_length = len(vocab)
+        return vocab
+        
+    def _get_glove_indices(self):
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: remove

File: dataset/jeopardy_dataset.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+    def _preprocess_sentence(self, sentence):
+        arr = nltk.word_tokenize(sentence)
+        arr = [word.lower() for word in arr]
+        if len(arr) > self.q_len:
+            return []
+        self._pad_arr(arr, self.q_len)
+        # TODO: look into START_TOKEN, END_TOKEN usefulness, perhaps ablation study?
+        return arr
+
+    def _pad_arr(self, arr, length):
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: remove

File: dataset/jeopardy_dataset.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        Returns the start index of the test split for the dataloader to use.
+        '''
+        return self.test_split_start
+    
+    def __len__(self):
+        return self.i
+
+    def __getitem__(self, idx):
+        # convert image id to image tensor on demand, not stored in RAM like that
+        question, image_id, answer = self.return_array[idx]
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: somewhre you should be calling `RobertaTokenizer`?

File: main.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+# output = model(**encoded_input)
+from torchvision.models import resnet18
+import torch.nn as nn
+import torch 
+from transformers.data.data_collator import DataCollatorForLanguageModeling
+
+class JeopardyModel(pl.LightningModule):
+    def __init__(self, config):
+        super().__init__()
+        self.tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
```
Comment (mhw32) [inline main.py]: Can you clean up the dataset class and move this to the dataset?

File: main.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        # TODO: freeze everything except the last 3 layers of Roberta
+        self.model = RobertaModel.from_pretrained('roberta-base')
+        self.word_embeddings = self.model.embeddings.word_embeddings
+        self.image_feature_extractor = resnet18(pretrained=False)
+        self.image_feature_extractor.fc = nn.Linear(512, self.word_embeddings.embedding_dim) 
+        self.alpha = config.mp.alpha
+        bs = config.mp.batch_size
+        self.idx = torch.tensor([bs-1] + [i for i in range(1, bs-1)])
+        sep_token = self.tokenizer(self.tokenizer.sep_token)
+        self.sep_token = self.word_embeddings([2], return_tensors='pt')
```
Comment (mhw32) [inline main.py]: fix: we should not be using word-embefdings

File: main.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      question, image1, image2, answer = batch
+      neg_image = image1.clone().view(-1)[self.idx].view(image1.size())
+      image1 = self.forward_image(image1)
+      image2 = self.forward_image(image2)
+
+      # TODO: efficient thing to do might be to use cached representations from previous forward
+      # passes for negatives?
+      neg_image = self.forward_image(neg_image) 
+
+      question_answer = self.tokenizer(question, answer, return_tensors='pt') # insert the SEP tokens - but not sure where padding comes in now
```
Comment (mhw32) [inline main.py:56]: FIx: i dont think this is right

**Review by mhw32 — CHANGES_REQUESTED** [review]
Incorrect usage of RobertaTokenizer. We should talk about this!

File: dataset/jeopardy_dataset.py (hunk 6 above)
Comment (shashank2000) [reply dataset/jeopardy_dataset.py:32]: can we talk about this in our meeting tomorrow

File: dataset/jeopardy_dataset.py (hunk 7 above)
Comment (shashank2000) [reply dataset/jeopardy_dataset.py]: pruned

File: dataset/jeopardy_dataset.py (hunk 8 above)
Comment (shashank2000) [reply dataset/jeopardy_dataset.py:67]: ah great point

File: dataset/jeopardy_dataset.py (hunk 6 above)
Comment (shashank2000) [reply dataset/jeopardy_dataset.py:32]: also might be worth it to get rid of q_len constraint entirely, I'll look at the frequency of various question lengths tomorrow

File: data_module.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+          self.questions_file, 
+          self.answers_file, 
+          self.coco_loc, 
+          self.test_transform, 
+          frequency_threshold=threshold,
+          train=False)
+        
+        torch.save(self.train_dataset, saved_train_file)
+        torch.save(self.test_dataset, saved_test_file)
+    breakpoint()
```
Comment (mhw32) [inline data_module.py]: remove breakpoint

Comment (mhw32) [reply data_module.py]: I'm hesitant to cache too much cuz it often introduces unexpected bugs. How long does this processing take?

File: data_module.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+      return DataLoader(self.train_dataset, batch_size=self.batch_size, shuffle=True,
+                        num_workers=self.num_workers, pin_memory=True, drop_last=True)
+
+  def test_dataloader(self):
+      return DataLoader(self.test_dataset, batch_size=self.batch_size,
+                        num_workers=self.num_workers, pin_memory=True, drop_last=True)
+
+  # TODO: fix the data caching stuff
+  @staticmethod
+  def file_name(multiple_images):
```
Comment (mhw32) [inline data_module.py:138]: what is this for?

File: dataset/jeopardy_dataset.py [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+        else:
+            self.questions = self.questions[split_index:]
+            self.answers = self.answers[split_index:]
+            
+        questions_dict = {q['question_id']: (q["question"], int(q["image_id"])) for q in self.questions}
+        self.images_dir = images_dir
+
+        self.image_id_to_filename = self._find_images()
+        self.transform = transform    
+        self.frequent_answers = self._find_frequent_answers(frequency_threshold)
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:81]: what is this for? add comments

File: dataset/jeopardy_dataset.py [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+        self.question_to_answer = self._find_answers()
+        self.tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
+
+        # filter down questions_dict based on the answer set we selected above
+        questions_dict = self.filter_questions_dict(questions_dict)
+        
+        self.return_array = [0] * len(questions_dict)
+        
+        self.i = 0
+        # figure out what to do with the attention mask
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: you need to return the mask and pass it roberta

File: dataset/jeopardy_dataset.py [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+        id_to_filename = {}
+        for filename in os.listdir(self.images_dir):
+            if not filename.endswith('.jpg'):
+                continue
+            id_and_extension = filename.split('_')[-1]
+            image_id = int(id_and_extension.split('.')[0])
+            id_to_filename[image_id] = filename
+        return id_to_filename
+
+    def get_split_index(self):
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: why do we need this?

File: main.py [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+    def shared_step(self, batch):
+      # recomputing the sep_token's embedding for each batch might be overkill
+      
+      question, image1, image2, answer = batch
+      # they are tokens, this is where we can mask a question token
+      q_labels = torch.ones_like(question['input_ids']) * -100
+      random_word = random.randint(0, len(question))
+      q_labels[random_word] = question[random_word]
+      breakpoint()
```
Comment (mhw32) [inline main.py]: remove

File: main.py [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+import pytorch_lightning as pl 
+from utils.model_utils import freeze_everything_except_top
+import random 
+
+class JeopardyModel(pl.LightningModule):
+    def __init__(self, config, pretrained_bert):
+        super().__init__()
+        # TODO: freeze everything except the last 3 layers of Roberta
+        self.model = RobertaForMaskedLM.from_pretrained('roberta-base')
+        freeze_everything_except_top(self.model, 3)
```
Comment (mhw32) [inline main.py]: nice! i like this utility.

File: main.py [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+        self.word_embeddings = self.model.roberta.embeddings.word_embeddings
+        self.image_feature_extractor = resnet18(pretrained=False)
+        self.image_feature_extractor.fc = nn.Linear(512, self.word_embeddings.embedding_dim) 
+        self.alpha = config.model_params.alpha
```
Comment (mhw32) [inline main.py:22]: what is alpha?

File: main.py [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+        self.sep_token = self.word_embeddings(torch.tensor([2]))
+        self.pooler = RobertaPooler(self.model.config)
+
+    def forward_image(self, x):
+      return self.image_feature_extractor(x)
+
+    def forward(self, x):
+      out = self.model.roberta(inputs_embeds=x, return_dict=True)
+      cls_embed = self.pooler(out)
+      return torch.linalg.norm(cls_embed, dim=-2)
```
Comment (mhw32) [inline main.py]: why normalize here?

File: main.py [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+      return loss 
+
+
+    def shared_step(self, batch):
+      # recomputing the sep_token's embedding for each batch might be overkill
+      
+      question, image1, image2, answer = batch
+      # they are tokens, this is where we can mask a question token
+      q_labels = torch.ones_like(question['input_ids']) * -100
+      random_word = random.randint(0, len(question))
```
Comment (mhw32) [inline main.py]: Careful! u need to make sure to mask all instances of this token. Use huggingface utilities to do this!

File: main.py [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+      sep_token = self.sep_token
+      breakpoint()
+      question_answer_image1 = torch.cat(question_answer, sep_token, image1, sep_token)
+      question_answer_image2 = torch.cat(question_answer, sep_token, image2, sep_token)
+      
+      neg_image = image1.view(image1.size())[self.idx]
+      question_answer_neg_image = question_answer + sep_token + neg_image + sep_token
+      contrastive_loss = self(question_answer_image1) + self(question_answer_image2) - self(question_answer_neg_image)
+      
+      mlm_loss = self.mlm_forward(question_answer_image1, labels=q_labels)
```
Comment (mhw32) [inline main.py]: doesn't huggingface mask out tokens for you already? 15% if i remember correctly

**Review by mhw32 — CHANGES_REQUESTED** [review]

File: data_module.py (hunk 18 above)
Comment (shashank2000) [reply data_module.py:138]: the name of the cache file

File: data_module.py (hunk 17 above)
Comment (shashank2000) [reply data_module.py]: like 5 minutes or so

File: main.py (hunk 24 above)
Comment (shashank2000) [reply main.py:22]: I thought we wanted the final loss to be a*contra_loss + (1-a)*mlm_loss

File: main.py (hunk 26 above)
Comment (shashank2000) [reply main.py]: done

File: main.py (hunk 27 above)
Comment (shashank2000) [reply main.py]: done

File: dataset/jeopardy_dataset.py (hunk 8 above)
Comment (shashank2000) [reply dataset/jeopardy_dataset.py:67]: this is fixed now btw, I randomize the questions when saving the file

File: .gitignore [hunk 28]
```diff
@@ -127,3 +128,7 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+
+# config file
+test.sh
+model.py
```
Comment (mhw32) [inline .gitignore:134]: should this be here?

File: config/v1model.json [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+        "q_len": 8
+    },
+    "optim_params": {
+        "batch_size": 8,
+        "learning_rate": 0.3,
+        "momentum": 0.9,
+        "weight_decay": 1e-6
+    },
+    "multiple_images": true,
+    "num_workers": 0,
```
Comment (mhw32) [inline config/v1model.json]: probably want to use more workers!

File: data_module.py [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+    def __call__(self, x):
+        sigma = random.uniform(self.sigma[0], self.sigma[1])
+        x = x.filter(ImageFilter.GaussianBlur(radius=sigma))
+        return x
+        
+class VQADataModule(LightningDataModule):
+  def __init__(self, batch_size, threshold=10, q_len=8, num_workers=8, val_split=0.2, num_answers=0, transfer=False, multiple_images=False):
+    super().__init__()
+
+    self.questions_file = self._make_shuffled_questions_file()
```
Comment (mhw32) [inline data_module.py]: add comment: what does `_make_shuffled_questions_file` do?

File: data_module.py (hunk 18 above)
Comment (mhw32) [reply data_module.py:138]: add comments please

File: data_module.py [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+    if multiple_images:
+      return "multiple"
+    return ""
+
+  def _make_shuffled_questions_file(self):
+    # if there is no shuffled questions file, make one
+    shuffled_questions_file = os.environ.get('SHUFFLED_QUESTIONS_FILE')
+    if not os.path.isfile(shuffled_questions_file):
+      original_questions_file = os.environ.get('QUESTIONS_FILE')
+      make_shuffled_questions_file(original_questions_file, shuffled_questions_file)
```
Comment (mhw32) [inline data_module.py]: what is this for? this looks like something we should do live.

File: dataset/jeopardy_dataset.py (hunk 8 above)
Comment (mhw32) [reply dataset/jeopardy_dataset.py:67]: but why are we randomizing questions? In general, I prefer to cache as few things as possible as it induces bugs very often.

File: dataset/jeopardy_dataset.py [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+            self.answers = self.answers[split_index:]
+            
+        questions_dict = {q['question_id']: (q["question"], int(q["image_id"])) for q in self.questions}
+        self.images_dir = images_dir
+
+        self.image_id_to_filename = self._find_images()
+        self.transform = transform    
+        self.frequent_answers = self._find_frequent_answers(frequency_threshold)
+        self.question_to_answer = self._find_answers()
+        self.tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:85]: add comments please

File: dataset/jeopardy_dataset.py [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+
+        # filter down questions_dict based on the answer set self.frequent_answers
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:87]: is this true? `frequent_answers` does not appear in the `filter_questions_dict` object

File: dataset/jeopardy_dataset.py [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+            answer set at the end of the day. We also filter the set of answers so we are dealing with only 
+            answer_confidence deals with the subject's confidence in answering the question, we are currently filtering
+            so its only yes's considered. I believe no and maybe are two other categories. 
+        '''
+        # TODO: more enlightened way of filtering answers. Check past VQA papers. 
+        word_freq = Counter([])
+        for i, ann in tqdm(enumerate(self.answers)):
+            for answer in ann['answers']:
+                actual_ans = answer['answer']
+                if only_one_word_answers and " " in actual_ans:
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: can we default `only_one_word_answers` to False! We shouldn't do that!

File: dataset/jeopardy_dataset.py [hunk 35]
```diff
@@ -0,0 +1,203 @@
...
+        return set([word for word in word_freq if word_freq[word] >= threshold])
+        
+
+    def _find_answers(self):
+        """
+            There several answers for each question in the dataset, for example, 
+            here's entry 0 of the answer vector for a random question:
+            {"answer": "net", "answer_confidence": "maybe", "answer_id": 1}
+            
+            For the sake of simplicity, we only consider answers that have answer_confidence: yes,
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: what do previous papers do? "answer => maybe" may be good for additional views?

Comment (mhw32) [reply dataset/jeopardy_dataset.py]: are there multiple answers with "yes" confidence? What do you do then?

File: dataset/jeopardy_dataset.py [hunk 36]
```diff
@@ -0,0 +1,203 @@
...
+        return self.i
+    
+    def _process_question(self, question_text):
+        tokenized_question = self.tokenizer(
+            question_text, return_tensors='pt', padding='max_length', truncation=True, max_length=self.q_len+2)
+        contrastive_input = tokenized_question["input_ids"].clone()
+        collator = DataCollatorForLanguageModeling(
+            self.tokenizer,
+            mlm=True,
+            mlm_probability=0.15,
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py]: make this a hyperparameter.

File: dataset/jeopardy_dataset.py [hunk 37]
```diff
@@ -0,0 +1,203 @@
...
+            answer_text, return_tensors='pt', padding='max_length', truncation=True, max_length=self.ans_len+2)
+        contrastive_input = tokenized_answer["input_ids"].clone()
+        collator = DataCollatorForLanguageModeling(
+            self.tokenizer,
+            mlm=True,
+            mlm_probability=0.15,
+        )
+        input_ids, labels = collator.mask_tokens(contrastive_input)
+        tokenized_answer['labels'] = labels
+        tokenized_answer['contrastive_input'] = contrastive_input
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:219]: what is this for? add label

File: dataset/jeopardy_dataset.py [hunk 38]
```diff
@@ -0,0 +1,203 @@
...
+    def _process_question(self, question_text):
+        tokenized_question = self.tokenizer(
+            question_text, return_tensors='pt', padding='max_length', truncation=True, max_length=self.q_len+2)
+        contrastive_input = tokenized_question["input_ids"].clone()
+        collator = DataCollatorForLanguageModeling(
+            self.tokenizer,
+            mlm=True,
+            mlm_probability=0.15,
+        )
+        input_ids, labels = collator.mask_tokens(contrastive_input)
```
Comment (mhw32) [inline dataset/jeopardy_dataset.py:195]: how does this interact with attention masks? Will `contrastive_input` and `input_ids` both have the same masks?

File: main.py (hunk 24 above)
Comment (mhw32) [reply main.py:22]: i see,add comments!

File: main.py [hunk 39]
```diff
@@ -0,0 +1,112 @@
...
+        self.sep_token = self.word_embeddings(torch.tensor([2])) # sep token is same as end token??
+        self.pooler = RobertaPooler(self.model.config) 
+
+    def forward_image(self, x):
+      return self.image_feature_extractor(x)
+
+    def forward(self, x):
+      out = self.model.roberta(**x)
+      # look into the input of pooler to see what it should be
+      cls_embed = self.pooler(out.last_hidden_state)
```
Comment (mhw32) [inline main.py:46]: what did you end up deciding on pooler? why have it out here?

File: main.py [hunk 40]
```diff
@@ -0,0 +1,112 @@
...
+      '''
+      q, img1, img2, a = batch
+      
+      img1 = self.forward_image(img1)
+      img2 = self.forward_image(img2)
+      img1 = torch.unsqueeze(img1, dim=1)
+      img2 = torch.unsqueeze(img2, dim=1)
+      
+      img_attention_mask = torch.ones_like(a["attention_mask"]) # sep + img + eos token
+      attention_mask = torch.cat((q["attention_mask"], a["attention_mask"], img_attention_mask), dim=2)
```
Comment (mhw32) [inline main.py]: does this take sep tokens into account?

File: main.py [hunk 41]
```diff
@@ -0,0 +1,112 @@
...
+
+      question = self.word_embeddings(q["contrastive_input"])
+      question = question.squeeze()
+      answer = answer.squeeze()
+      
+      '''Contrastive Loss'''
+      question_answer_image1 = torch.cat((question, answer, sep_token, img1, sep_token), dim=1)
+      question_answer_image2 = torch.cat((question, answer, sep_token, img2, sep_token), dim=1)
+      
+      neg_image = img1.view(img1.size())[self.idx] # verify this is really shuffling stuff
```
Comment (mhw32) [inline main.py]: we should walk through this

File: main.py [hunk 42]
```diff
@@ -0,0 +1,112 @@
...
+      '''
+        some notes:
+        - sep_token is the same as the eos_token in huggingface
+      '''
+      q, img1, img2, a = batch
+      
+      img1 = self.forward_image(img1)
+      img2 = self.forward_image(img2)
+      img1 = torch.unsqueeze(img1, dim=1)
+      img2 = torch.unsqueeze(img2, dim=1)
```
Comment (mhw32) [inline main.py:73]: unsqueeze to pretend to be a token?

File: main.py [hunk 43]
```diff
@@ -0,0 +1,112 @@
...
+      
+      img_attention_mask = torch.ones_like(a["attention_mask"]) # sep + img + eos token
+      attention_mask = torch.cat((q["attention_mask"], a["attention_mask"], img_attention_mask), dim=2)
+      
+      answer = self.word_embeddings(a["contrastive_input"])
+      sep_token = answer[:, :, -1]
```
Comment (mhw32) [inline main.py]: why is this the sep_token?

File: main.py [hunk 44]
```diff
@@ -0,0 +1,112 @@
...
+      
+      qai1_dict = self.make_dict(question_answer_image1, attention_mask)
+      qai2_dict = self.make_dict(question_answer_image2, attention_mask)
+      qaineg_dict = self.make_dict(question_answer_neg_image, attention_mask)
+
+      trip1 = self(qai1_dict)
+      trip2 = self(qai2_dict)
+      tripneg = self(qaineg_dict)
+      
+      contrastive_loss =  torch.linalg.norm(trip1.T@tripneg) - torch.linalg.norm(trip1.T@trip2) # could just use simclr class instead
```
Comment (mhw32) [inline main.py]: i dont think this is right! lets talk about it

File: main.py [hunk 45]
```diff
@@ -0,0 +1,112 @@
...
+      question = question.squeeze()
+
+      answer = self.word_embeddings(a["input_ids"])
+      answer = answer.squeeze()
+      question_answer_image_mlm = torch.cat((question, answer, sep_token, img1, sep_token), dim=1)
+
+      qai_mlm_dict = self.make_dict(question_answer_image_mlm, attention_mask)
+      # loss is only calculated on masked tokens in any case
+      labels = torch.zeros_like(attention_mask)
+      labels[:, :, :10] = q["labels"]
```
Comment (mhw32) [inline main.py]: why `:10`?

File: main.py [hunk 46]
```diff
@@ -0,0 +1,112 @@
...
+      qai_mlm_dict["labels"] = labels
+      mlm_loss = self.mlm_forward(qai_mlm_dict)
+      # grad can only be created implicitly for scalar outputs
+      return self.alpha*mlm_loss * (1-self.alpha)*contrastive_loss
+
+    def configure_optimizers(self):
+      # positional encodings for the image as well
+      return torch.optim.Adam(self.parameters(), lr=self.op.learning_rate)
```
Comment (mhw32) [inline main.py]: usually is `AdamW`... use very specific nlp

**Review by mhw32 — CHANGES_REQUESTED** [review]

File: main.py (hunk 44 above)
Comment (mhw32) [reply main.py]: torch.sum(trip1 * trip2) / tau - torch.sum(trip * tripneg) / tau

File: main.py (hunk 46 above)
Comment (mhw32) [reply main.py]:
```
            optimizer = torch.optim.AdamW(
                self._all_parameters(),
                lr=self.config.optim.learning_rate,
                betas=(0.9, 0.999),
                eps=1e-6,
                weight_decay=0.01,
            )
            num_training_steps = len(self.train_dataset) * self.config.optim.num_epochs
            scheduler = get_linear_schedule_with_warmup(
                optimizer,
                num_warmup_steps=self.config.optim.warmup_steps,
                num_training_steps=num_training_steps,
            )
```

Comment (mhw32) [reply main.py]: from transformers import get_linear_schedule_with_warmup

Comment (mhw32) [reply main.py]:
```
        "learning_rate": 1e-4,
        "weight_decay": 0,
        "patience": 1000,
        "num_epochs": 400,
        "warmup_steps": 10000,
        "use_scheduler": true
        ```

File: dataset/jeopardy_dataset.py (hunk 38 above)
Comment (shashank2000) [reply dataset/jeopardy_dataset.py:195]: I don't recall what we talked about wrt this - the answer to your question is yes though - same attention mask for MLM and contrastive loss. Is this not what I was supposed to do?

File: main.py (hunk 43 above)
Comment (shashank2000) [reply main.py]: could we talk about this when we meet? I think I'm being inefficient on memory usage here

**Pull request merged**